### PR TITLE
Bug 1658349 - Allow using the quantity metric type outside of Gecko

### PR DESCRIPTION
**Pull request by brizental**
glean_parser PR: https://github.com/mozilla/glean_parser/pull/225

### CODE REVIEW

**Comment by brizental** [comment]
After #1200 I should probably drop the C# commit, but then it would have glean_parser create invalid metrics.

**Comment by Dexterp37** [comment]
> 
> 
> After #1200 I should probably drop the C# commit, but then it would have glean_parser create invalid metrics.

Nah, it's fine to leave this in. No worries.

File: glean-core/csharp/Glean/Metrics/QuantityMetricType.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+            Dispatchers.LaunchAPI(() => {
+                SetSync(value);
+            });
+        }
+
+        /// <summary>
+        /// Internal only, synchronous API for setting a quantity value.
+        /// </summary>
+        /// <param name="value">The value to set. Must be non-negative.</param>
+        internal void SetSync(Int32 value)
```
Comment (Dexterp37) [inline glean-core/csharp/Glean/Metrics/QuantityMetricType.cs:76]: Do we need this? We're not really using it anywhere.

File: glean-core/ios/Glean/Metrics/QuantityMetric.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import Foundation
+
+/// This implements the developer facing API for recording counter metrics.
+///
+/// Instances of this class type are automatically generated by the parsers at build time,
+/// allowing developers to record values that were previously registered in the metrics.yaml file.
+///
+/// The counter API only exposes the `QuantityMetricType.set(_:)` method, which takes care of validating the input
+/// data and making sure that requirements are enforced.
+public class QuantityMetricType {
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Metrics/QuantityMetric.swift:14]: @badboy would you kindly review the Swift parts?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]
Looks generally solid. A couple of comments, below.

File: glean-core/csharp/Glean/Metrics/QuantityMetricType.cs (hunk 1 above)
Comment (brizental) [reply glean-core/csharp/Glean/Metrics/QuantityMetricType.cs:76]: We use it in line 68, if you mean the `SetSync`. Or do you mean the whole C# part?

Comment (brizental) [reply glean-core/csharp/Glean/Metrics/QuantityMetricType.cs:76]: Just saw your comment about not droppping the C# part.

Comment (Dexterp37) [reply glean-core/csharp/Glean/Metrics/QuantityMetricType.cs:76]: Yes, but we just use it there. We have the "sync" functions to make it possibly to synchronously set metrics during the Glean  init. But, for this case, we don't need it. Let's move the content of this function to the only call site, please! :)

**Review by Dexterp37 — APPROVED** [review]
r+. Let's not merge until somebody reviews the Swift part, though

File: glean-core/ios/Glean/Metrics/QuantityMetric.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+
+/// This implements the developer facing API for recording counter metrics.
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/QuantityMetric.swift:7]:
```suggestion
/// This implements the developer facing API for recording quantity metrics.
```

File: glean-core/ios/Glean/Metrics/QuantityMetric.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+///
+/// Instances of this class type are automatically generated by the parsers at build time,
+/// allowing developers to record values that were previously registered in the metrics.yaml file.
+///
+/// The counter API only exposes the `QuantityMetricType.set(_:)` method, which takes care of validating the input
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/QuantityMetric.swift:12]:
```suggestion
/// The quantity API only exposes the `QuantityMetricType.set(_:)` method, which takes care of validating the input
```

File: glean-core/ios/Glean/Metrics/QuantityMetric.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    ///
+    /// - parameters:
+    ///     * pingName: represents the name of the ping to retrieve the metric for.
+    ///                 Defaults to the first value in `sendInPings`.
+    /// - returns: true if metric value exists, otherwise false
+    public func testHasValue(_ pingName: String? = nil) -> Bool {
+        Dispatchers.shared.assertInTestingMode()
+
+        let pingName = pingName ?? self.sendInPings[0]
+        return glean_quantity_test_has_value(self.handle, pingName) != 0
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/QuantityMetric.swift:73]: We should probably start with using `.toBool()` for this everywhere.
Right now the `!= 0` check is more common, so I'm good with keeping it here and we change that in one swoop across all metrics in a follow-up.

**Review by badboy — CHANGES_REQUESTED** [review]
This requires documentation updates here: https://mozilla.github.io/glean/book/user/metrics/quantity.html

Please also add a changelog entry.

File: glean-core/ios/Glean/Metrics/QuantityMetric.swift (hunk 5 above)
Comment (brizental) [reply glean-core/ios/Glean/Metrics/QuantityMetric.swift:73]: https://bugzilla.mozilla.org/show_bug.cgi?id=1663469

File: docs/user/metrics/quantity.md [hunk 6]
```diff
@@ -1,53 +1,152 @@
 # Quantity
 
-Used to record a single non-negative integer value.
+Used to record a single non-negative integer value or 0.
 For example, the width of the display in pixels.
 
-> **Note**: Quantities are currently only allowed for GeckoView metrics (the `gecko_datapoint` parameter is present) and thus have only a Kotlin API.
+> **IMPORTANT** If you need to _count_ something (e.g. number of tabs open or number of times a button is pressed) prefer using the [Counter](./counter.md) metric type, which has a specific API for counting things and also takes care of resetting the count at the correct time.
```
Comment (badboy) [inline docs/user/metrics/quantity.md:6]: I think this is good, but cc @Dexterp37 if he agrees.

File: docs/user/metrics/quantity.md [hunk 7]
```diff
@@ -1,53 +1,152 @@
 # Quantity
 
-Used to record a single non-negative integer value.
+Used to record a single non-negative integer value or 0.
```
Comment (badboy) [inline docs/user/metrics/quantity.md:3]: *math head on* weeeeell, `non-negative` means including 0 :D 
Maybe:

```suggestion
Used to record a single non-negative integer value (including 0).
```

File: docs/user/metrics/quantity.md [hunk 8]
```diff
@@ -1,53 +1,152 @@
...
+```
+
+</div>
+
+{{#include ../../tab_footer.md}}
+
+## Limits
 
-* Quantities are only available for metrics that come from Gecko.
+* Quantities must be non-negative integers or 0.
```
Comment (badboy) [inline docs/user/metrics/quantity.md:149]:
```suggestion
* Quantities must be non-negative integers (including 0).
```

**Review by badboy — APPROVED** [review]
Last couple of comments but then good to go.

File: docs/user/metrics/quantity.md (hunk 7 above)
Comment (brizental) [reply docs/user/metrics/quantity.md:3]: TIL!

File: docs/user/metrics/quantity.md (hunk 6 above)
Comment (Dexterp37) [reply docs/user/metrics/quantity.md:6]: Looks good to me! Sorry for the delay!

**Pull request merged**